terrain/terrain.cc

```diff
@@ -0,0 +1,530 @@
+
+
+    float *vertexBuffer = (float*)malloc(maxVertexCount * 3 * sizeof(float));
+    float *normalBuffer = (float*)malloc(maxVertexCount * 3 * sizeof(float));
+    int *indexBuffer = (int*)malloc(maxIndexCount * sizeof(int));
+    int *chunkVertexRangeBuffer = (int*)malloc(chunkCount * chunkCount * 2 * sizeof(int));
+    int *vertexFreeRangeBuffer = (int*)malloc(chunkCount * chunkCount * 2 * sizeof(int));
+    int *chunkIndexRangeBuffer = (int*)malloc(chunkCount * chunkCount * 2 * sizeof(int));
+    int *indexFreeRangeBuffer = (int*)malloc(chunkCount * chunkCount * 2 * sizeof(int));
+
+    // std::vector<int> chunkGroup = {};
+
+    int totalVertexCount = 0;
+    int totalIndexCount = 0;
+
+    for (int i = 0; i < chunkCount; i++) {
+        for (int j = 0; j < chunkCount; j++) {
+            float origin[3] = {
+                (float)(i - chunkCount / 2) * chunkSize,
+                0.0,
+                (float)(j - chunkCount / 2) * chunkSize
+            };
+
+            // chunkGroup.push_back(groupStart);
+            int chunkIndex = i * chunkCount + j;
+
+            std::vector<Vector3> vertices = {};
+            std::vector<Vector3> normals = {};
+            std::vector<int> indices = {};
+
+            // int groupStart = totalIndexCount;
+
+            createChunk(
+                origin,
+                chunkSize,
+                segment,
+                vertices,
+                normals,
+                indices
+            );
+
+            for (int i = 0; i < indices.size(); i++) {
+                indices[i] += totalVertexCount;
+            }
+
+            memcpy(vertexBuffer + totalVertexCount * 3, &(vertices.front()), vertices.size() * sizeof(Vector3));
+            memcpy(normalBuffer + totalVertexCount * 3, &(normals.front()), normals.size() * sizeof(Vector3));
+            memcpy(indexBuffer + totalIndexCount, &(indices.front()), indices.size() * sizeof(int));
+
+            chunkVertexRangeBuffer[chunkIndex * 2] = totalVertexCount;
+            chunkVertexRangeBuffer[chunkIndex * 2 + 1] = vertices.size();
+
+            chunkIndexRangeBuffer[chunkIndex * 2] = totalIndexCount;
+            chunkIndexRangeBuffer[chunkIndex * 2 + 1] = indices.size();
+
+            totalVertexCount += vertices.size();
+            totalIndexCount += indices.size();
+
+            // chunkGroup.push_back(indices.size() - groupStart);
+        }
+    }
+
+    vertexFreeRangeBuffer[0] = totalVertexCount;
+    vertexFreeRangeBuffer[1] = maxVertexCount - totalVertexCount;
+
+    indexFreeRangeBuffer[0] = totalIndexCount;
+    indexFreeRangeBuffer[1] = maxIndexCount - totalIndexCount;
+
+    for (int i = 1; i < chunkCount * chunkCount; i++) {
+        vertexFreeRangeBuffer[2 * i] = 0;
+        vertexFreeRangeBuffer[2 * i + 1] = 0;
+        indexFreeRangeBuffer[2 * i] = 0;
+        indexFreeRangeBuffer[2 * i + 1] = 0;
+    }
+
+    int *outputBuffer = (int*)malloc(7 * sizeof(int));
```

Comment:
It's not really clear why this has `7`. This should be commented.

Reply:
Alternatively, this could be `sizeof(type) + ...` to make it clear at least the structure of the data expected. But we should avoid magic numbers with no justification.

Reply:
We are returning 7 buffers. Positions, normals, indices and so on

Reply:
And the output buffer just contains addresses of those buffers